hibernate-rx-core/src/main/java/org/hibernate/rx/event/impl/DefaultRxLoadEventListener.java

```diff
@@ -74,27 +76,31 @@ public void onLoad(
 			throw new HibernateException( "Unable to locate persister: " + event.getEntityClassName() );
 		}
 
-		Object checkId = checkId( event, loadType, persister );
-		if ( checkId instanceof CompletionStage) {
+		CompletionStage<Void> checkId = checkId( event, loadType, persister );
+		if ( !checkId.toCompletableFuture().isDone() ) {
 			// This only happens if the object is loaded from the db
-			throw new UnsupportedOperationException( "Unexpected access to the db" );
+			throw new AssertionFailure( "Unexpected access to the database" );
 		}
 
 		try {
-			Object loaded = doOnLoad( persister, event, loadType );
-			if ( loaded instanceof CompletionStage ) {
+			CompletionStage<Optional<Object>> loaded = doOnLoad( persister, event, loadType );
+			if ( !loaded.toCompletableFuture().isDone() ) {
```

Comment:
Out of curiosity, this would work if the db is super quick and the stage completes before the check. isn't it?

Reply:
Assuming there is no bug in our code, this should never be called on a `CompletionStage` coming from the DB. If it *is*, the worst that can possibly happen is that we don't notice the bug in our code.